airbyte-integrations/connectors/source-railz-ai/source_railz_ai/components.py

```diff
@@ -0,0 +1,370 @@
+        self._options = options
+        self._cursor = {}
+
+    def get_stream_state(self) -> StreamState:
+        return self._cursor if self._cursor else {}
+
+    def update_cursor(self, stream_slice: StreamSlice, last_record: Optional[Record] = None):
+        if last_record is None:
+            self._cursor.update(stream_slice)
+            return
+
+        if stream_slice["businessName"] not in self._cursor:
+            self._cursor[stream_slice["businessName"]] = set()
+        self._cursor[stream_slice["businessName"]].add(stream_slice["serviceName"])
+
+    @staticmethod
+    def _check_valid_service(connection):
+        return connection["status"] in ["active", "disconnected", "expired"]
+
+    def stream_slices(self, sync_mode: SyncMode, stream_state: StreamState) -> Iterable[StreamSlice]:
+        if not self.parent_stream_config:
+            yield from []
+        else:
+            parent_stream = self.parent_stream_config.stream
+            for parent_stream_slice in parent_stream.stream_slices(sync_mode=sync_mode, cursor_field=None, stream_state=stream_state):
+                empty_parent_slice = True
+
+                for parent_record in parent_stream.read_records(
+                    sync_mode=SyncMode.full_refresh, cursor_field=None, stream_slice=parent_stream_slice, stream_state=None
+                ):
+                    # Skip non-records (eg AirbyteLogMessage)
+                    if isinstance(parent_record, AirbyteMessage):
+                        if parent_record.type == Type.RECORD:
+                            parent_record = parent_record.record.data
+                        else:
+                            continue
+
+                    for connection in parent_record["connections"]:
+                        if self._check_valid_service(connection):
+                            empty_parent_slice = False
+                            yield {"businessName": parent_record["businessName"], "serviceName": connection["serviceName"]}
+
+                if empty_parent_slice:
+                    yield from []
+
+    def get_request_params(
+        self,
+        stream_state: Optional[StreamState] = None,
+        stream_slice: Optional[StreamSlice] = None,
+        next_page_token: Optional[Mapping[str, Any]] = None,
+    ) -> Mapping[str, Any]:
+        options = {}
+        if stream_slice:
+            options = {
+                "businessName": stream_slice["businessName"],
+                "serviceName": stream_slice["serviceName"],
+            }
+        return options
+
+
+@dataclass
+class RailzAiDatetimeStreamSlicer(DatetimeStreamSlicer):
+    cursor_format: str = '%Y-%m-%d'
+
+    def parse_date(self, date: str) -> datetime:
+        return self._parser.parse(date, self.cursor_format, self._timezone)
+
+
+@dataclass
+class RailzAiIncrementalReportsSlicer(RailzAiDatetimeStreamSlicer):
+    def update_cursor(self, stream_slice: StreamSlice, last_record: Optional[Record] = None):
```

Comment:
it's not super clear to me how this differs from the builtin `DatetimeStreamSlicer`. is this to work around a bug in the parent class or implementing a new behavior specific to railz.ai?